File: 0014_longest_common_prefix/trie/go/solution.go
package trie

type TrieNode struct {
	children   map[rune]*TrieNode
	isEnd      bool
	childCount int
}

type Trie struct {
	root *TrieNode
}

func NewTrie() *Trie {
	return &Trie{root: &TrieNode{children: make(map[rune]*TrieNode)}}
}

func (t *Trie) Insert(word string) {
	node := t.root
	for _, char := range word {
		if _, ok := node.children[char]; !ok {
			node.children[char] = &TrieNode{children: make(map[rune]*TrieNode)}
			node.childCount++
		}
		node = node.children[char]
	}
	node.isEnd = true
}

func (t *Trie) FindLongestPrefix() string {
	node := t.root
	prefix := ""
	for node.childCount == 1 && !node.isEnd {
		for char, child := range node.children {
			prefix += string(char)
			node = child
			break
		}
	}
	return prefix
}

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}
	if len(strs) == 1 {
		return strs[0]
	}

	trie := NewTrie()
	for _, s := range strs {
		if s == "" {
			return ""
		}
		trie.Insert(s)
	}
	return trie.FindLongestPrefix()
}

File: 0014_longest_common_prefix/trie/kotlin/solution.kt
private class TrieNode {
    val children = mutableMapOf<Char, TrieNode>()
    var isEnd = false
}

private class Trie {
    private val root = TrieNode()

    fun insert(word: String) {
        var node = root
        for (char in word) {
            node = node.children.getOrPut(char) { TrieNode() }
        }
        node.isEnd = true
    }

    fun findLongestPrefix(): String {
        val prefix = StringBuilder()
        var node = root

        while (node.children.size == 1 && !node.isEnd) {
            val (char, next) = node.children.entries.first()
            prefix.append(char)
            node = next
        }

        return prefix.toString()
    }
}

class Solution {
    fun longestCommonPrefix(strs: Array<String>): String {
        if (strs.isEmpty()) return ""
        if (strs.size == 1) return strs[0]

        val trie = Trie()
        for (str in strs) {
            if (str.isEmpty()) return ""
            trie.insert(str)
        }

        return trie.findLongestPrefix()
    }
}
